### PR TITLE
pythonPackages.poetry: init at 0.12.10

### LINKED ISSUE #53132: [New Package Request] Add Poetry – for Linux et al.
Python's [greatest](https://poetry.eustace.io/) packaging and dependency management tool.

Thanks in advance!

### PR DESCRIPTION
**Pull request by jbaum98**
###### Motivation for this change

Add the `poetry` package and its dependencies.

Closes #53132.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/clikit/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/clikit/default.nix]: Coding conventions say when this doesn't fit on one line it should be written as
```
{ lib , buildPythonPackages, fetchPypi
, isPy27, ...
```

> https://nixos.org/nixpkgs/manual/#sec-syntax

File: pkgs/development/python-modules/clikit/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/clikit/default.nix]: Probably a copy error here, see https://github.com/sdispater/clikit

File: pkgs/development/python-modules/cleo/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cleo/default.nix]: Same as https://github.com/NixOS/nixpkgs/pull/53599/commits/045474ca7a1177e30ba55e9e7435d233beb7044e#r245821465

File: pkgs/development/python-modules/cleo/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cleo/default.nix]:
```suggestion
    description = "Allows you to create beautiful and testable command-line interfaces";
```

We don't like to have the package name in the `description`.

File: pkgs/development/python-modules/pre-commit/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/pre-commit/default.nix]: This is already packaged as an application https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/version-management/git-and-tools/pre-commit/default.nix
Though this has some nice improvements.

Since you've went ahead and packaged this as a library and it already exists as an application, perhaps we should use `toPythonApplication` for `pre-commit`? See the nixpkgs manual for explanation.

If not I would go ahead and merge these changes somehow into the current `pre-commit` derivation.

Comment (worldofpeace) [inline pkgs/development/python-modules/pre-commit/default.nix]: This is already packaged as an application https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/version-management/git-and-tools/pre-commit/default.nix
Though this has some nice improvements.

Why was this added as a python-module?

File: pkgs/development/python-modules/pylev/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/pylev/default.nix]: I'm not seeing the test in pypi tarball. I don't think we should go through any trouble to get it though.

So the `checkPhase` should be disabled with

```nix
doCheck = false;
```

File: pkgs/development/python-modules/pastel/default.nix
Comment (worldofpeace) [reply pkgs/development/python-modules/pastel/default.nix]: Also this won't build because we're missing `pytest` in `checkInputs`.

Comment (worldofpeace) [reply pkgs/development/python-modules/pastel/default.nix]: Actually, I don't see any tests in the tarball...

File: pkgs/development/python-modules/clikit/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/clikit/default.nix]: I'm also not seeing any tests in this tarball also. So checks should be disabled.

File: pkgs/development/python-modules/cleo/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cleo/default.nix]: No tests in this tarball either :frowning:

File: pkgs/development/python-modules/cachy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cachy/default.nix]: :rofl: No tests in here either, don't get why people aren't distributing these things.

I'm flipping the switch and saying that we should probably use `fetchFromGitHub` to get a source archive for all packages that don't have tests on pypi. Only exception would be when the tests look problematic.

**Comment by worldofpeace** [comment]
@jbaum98 Why `enableParallelBuilding = true;` in all the packages you've added?

**Comment by jbaum98** [comment]
Thanks for your thorough feedback and sorry for what appears to have been a pretty sloppy PR.

> @jbaum98 Why `enableParallelBuilding = true;` in all the packages you've added?

This could be entirely misguided, but I always set `enableParallelBuilding = true;` because I think 99% you want to be using all your cores, in part based on #5489.

File: pkgs/development/python-modules/pastel/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/pastel/default.nix]: I can answer the `pytest` part of the mystery: it appears to be downloading test dependencies by itself:
```
running install tests
running test
Searching for pytest
Reading https://pypi.org/simple/pytest/
Downloading https://files.pythonhosted.org/packages/95/6a/4122affe57eb3857874ef959a73f362e51b23812b96c92e073b16e1effd0/pytest-4.1.0-py2.py3-none-any.whl#sha256=3e65a22eb0d4f1bdbc1eacccf4a3198bf8d4049dea5112d70a0c61b00e748d02
Best match: pytest 4.1.0
Processing pytest-4.1.0-py2.py3-none-any.whl
Installing pytest-4.1.0-py2.py3-none-any.whl to /tmp/nix-build-python3.7-pastel-0.1.0.drv-0/pastel-0.1.0/.eggs
writing requirements to /tmp/nix-build-python3.7-pastel-0.1.0.drv-0/pastel-0.1.0/.eggs/pytest-4.1.0-py3.7.egg/EGG-INFO/requires.txt
```

But not sure where it's getting the tests themselves, or the dependencies.

File: pkgs/development/python-modules/clikit/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/clikit/default.nix]: Although for `pylev` getting the GitHub source worked so we could run tests because it uses `setup.py`, for `clikit` I'm running into a chicken-egg problem because it wants to use `poetry` as its build system to build from `pyproject.toml`. Thoughts?

My instinct is we bootstrap just by not using tests, and then somehow use that that `poetry` to rebuild its dependencies with tests, but I'm not sure how to do that.

File: pkgs/development/python-modules/cleo/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/cleo/default.nix]: Also wants to build using `poetry`.

File: pkgs/development/python-modules/cachy/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/cachy/default.nix]: Also want's `poetry`.

File: pkgs/development/python-modules/tomlkit/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/tomlkit/default.nix]: Also wants poetry.

File: pkgs/development/python-modules/pastel/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/pastel/default.nix]: Nevermind, it's using `tox` to download the test dependencies. Switching to GitHub and specifying the test dependencies and test script myself fixes everything, and all problems with Python 3.

**Comment by worldofpeace** [comment]
> Although for pylev getting the GitHub source worked so we could run tests because it uses setup.py, for clikit I'm running into a chicken-egg problem because it wants to use poetry as its build system to build from pyproject.toml. Thoughts?
>
>My instinct is we bootstrap just by not using tests, and then somehow use that that poetry to rebuild its dependencies with tests, but I'm not sure how to do that.

I think we'd have to add a `build-python-package-poetry` at https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/interpreters/python if we wanted to build these project that use poetry. And obviously we'd have to merge this pr first before that.

However it is confusing that they'd distribute a `setup.py` in the tarball but not on their github repos.

File: pkgs/development/python-modules/pre-commit/default.nix
Comment (worldofpeace) [reply pkgs/development/python-modules/pre-commit/default.nix]: All is not lost, I will *borrow* some of these changes and incorporate them into nixpkgs somehow :smile:

File: pkgs/development/python-modules/tomlkit/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/tomlkit/default.nix]: I think I fixed that by actually adding the dependencies.

**Comment by jbaum98** [comment]
> However it is confusing that they'd distribute a `setup.py` in the tarball but not on their github repos.

I'm guessing `poetry` produces the `setup.py` automatically when distributing because `poetry` isn't ubiquitous.

File: pkgs/development/python-modules/tomlkit/default.nix
Comment (worldofpeace) [reply pkgs/development/python-modules/tomlkit/default.nix]: That'd probably do it.

**Comment by jbaum98** [comment]
Okay, I think I've fixed anything that isn't the `poetry` package test situation. Should I squash these commits one per package or one all together and we can land this PR, and then figure out how to bootstrap poetry?

**Comment by worldofpeace** [comment]
> Okay, I think I've fixed anything that isn't the `poetry` package test situation. Should I squash these commits one per package or one all together and we can land this PR, and then figure out how to bootstrap poetry?

I'll give your new commits a quick look to see if they're resolved properly. And yes we'd like a commit for each package.

As for the poetry test situation, I agree with that flow.

**Comment by worldofpeace** [comment]
Remember to add `doCheck = false;` for the ones that we still can't get tests for.
Also add a comment above with why.

**Comment by worldofpeace** [comment]
@jbaum98 Great, squash at will. Then I will begin some testing locally.

**Comment by jbaum98** [comment]
@basvandijk @Infinisil @peti @ryantm Sorry please ignore the review, I accidentally did something wrong while squashing and included your commits temporarily.

**Comment by worldofpeace** [comment]
Just got informed that `enableParallelBuilding = true;` probably does nothing unless you're using `format = other`.  So basically using an entirely different 'build system' where parallelization would be nice.

**Comment by jbaum98** [comment]
Hmm, I've never heard of `format = other`. What does it do?

I just did a quick experiment and by adding `enableParallelBuilding = true;` to `hello`, the build flags changed from
```
build flags: SHELL=/nix/store/gayzik3jx16rpc64czk5cy5s953431s6-bash-4.4-p23/bin/bash
```
to
```
build flags: -j4 -l4 SHELL=/nix/store/gayzik3jx16rpc64czk5cy5s953431s6-bash-4.4-p23/bin/bash
```
which is pretty much what I'm going for.

Do you mean that it doesn't enable package level parallelism?

File: pkgs/development/python-modules/poetry/jsonschema.nix
Comment (jbaum98) [reply pkgs/development/python-modules/poetry/jsonschema.nix]: I was translating from [here](https://github.com/Julian/jsonschema/blob/8cc6a5af0b5ab343707c0c71021c477651aa479b/setup.cfg#L30), and I guess it makes a difference if we ever had multiple supported python 2 versions. But if you think it looks weird I can change it.

Comment (worldofpeace) [reply pkgs/development/python-modules/poetry/jsonschema.nix]: I feel like it's unlikely that this would happen for python 2.

**Comment by worldofpeace** [comment]
> I just did a quick experiment and by adding enableParallelBuilding = true; to hello, the build flags changed from
>
> `build flags: SHELL=/nix/store/gayzik3jx16rpc64czk5cy5s953431s6-bash-4.4-p23/bin/bash`
>
>to
>
>`build flags: -j4 -l4 SHELL=/nix/store/gayzik3jx16rpc64czk5cy5s953431s6-bash-4.4-p23/bin/bash`
>
> which is pretty much what I'm going for.

Actually I think that's all it does https://github.com/NixOS/nixpkgs/blob/e39ff06b16324025ecfe50e70b890e2286ff4338/pkgs/stdenv/generic/setup.sh#L1021

Though I'm only seeing like  17 occurrences of this being enabled for python-modules.
So it's not widely considered as being useful.

Btw, currently a dependent of `poetry` is broken `pyrsistent`.
Here's a log from [hydra](https://nix-cache.s3.amazonaws.com/log/2y57fcfp9b3pd7mk73lpcd57c3av3mm2-python3.7-pyrsistent-0.14.6.drv), this can be fixed here.

**Comment by jbaum98** [comment]
I think that the build has been broken for a month or two: https://hydra.nixos.org/build/86609521#tabs-summary.
And it seems like Hydra is having trouble reaching pypi.org.

**Comment by worldofpeace** [comment]
Yep here's some graphs  of the latest builds of `pyrsistent` on master
* https://hydra.nixos.org/job/nixpkgs/trunk/python37Packages.pyrsistent.x86_64-linux
* https://hydra.nixos.org/job/nixpkgs/trunk/python27Packages.pyrsistent.x86_64-linux

I have a fix incoming to your branch if you don't mind.

**Comment by jbaum98** [comment]
Go for it. I realized it's something to do with sandboxed builds, got it to fail locally.

**Comment by worldofpeace** [comment]
> Go for it. I realized it's something to do with sandboxed builds, got it to fail locally.

Yeah, only networking allowed in a sandboxed build is with yourself.

**Comment by jbaum98** [comment]
Why do you use `nativeBuildInputs` here and not for `pytest`?

**Comment by worldofpeace** [comment]
Because `pytestrunner` is only needed for the setup script and `nativeBuildInputs` are usually for stuff  you need to run as part of the build.

Though I should probably move `pytest` to `checkInputs` which only get added to `buildInputs` when `doCheck = true;`.

**Comment by jbaum98** [comment]
But shouldn't `checkInputs` get added to `nativeBuildInputs` and not `buildInputs`, because you only need them when building?

**Comment by worldofpeace** [comment]
We're talking about changing that in #53440

**Comment by worldofpeace** [comment]
Also speaking of `setup_requires`, `jsonschema` needs `setuptools_scm`.
https://github.com/Julian/jsonschema/blob/37671acda2424f1bb13a5e2f5e8f81441d73487b/setup.cfg#L26

**Comment by jbaum98** [comment]
Ah I got it to fail with sandbox. Nice catch.

**Comment by worldofpeace** [comment]
Everything builds and `poetry` executes, though I haven't tested its function.

One weird thing I did notice is that the help menu is showing an absolute path to `poetry`.
```
Help:
 The help command displays help for a given command:
 
   python /nix/store/j5lbxb7g1xg210axcjsiw7l7zlm2f2id-python3.7-poetry-0.12.10/bin/.poetry-wrapped help list
 
 You can also output the help in other formats by using the --format option:
 
   python /nix/store/j5lbxb7g1xg210axcjsiw7l7zlm2f2id-python3.7-poetry-0.12.10/bin/.poetry-wrapped help --format=json list
 
 To display the list of available commands, please use the list command.
```

Tried setting `--argv0` to `poetry` as a hopeful fix, but to no avail.

File: pkgs/development/python-modules/cachy/default.nix
Comment (worldofpeace) [inline pkgs/development/python-modules/cachy/default.nix]: *nitpick warning*

Can you add a line of whitespace here, and wherever it isn't present in your derivations?

**Review by worldofpeace — APPROVED** [review]
I successfully produced a wheel with `poetry build --format wheel`.

Though I'd like another set of eyes to look this over before merging.
And see how they feel about `enableParallelBuilding` here.
cc @Ma27 perhaps?

**Comment by worldofpeace** [comment]
As for being able to build poetry based python packages, I think we can do something similar to what's done for [flit](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/build-python-package-flit.nix).

File: pkgs/development/python-modules/clikit/default.nix
Comment (Ma27) [inline pkgs/development/python-modules/clikit/default.nix]: we don't have Python 3.4 on master anymore, so this shouldn't be needed.

File: pkgs/development/python-modules/poetry/default.nix
Comment (Ma27) [inline pkgs/development/python-modules/poetry/default.nix]: please indent properly here. Unfortunately I don't know of a proper autofixing tool for Nix CS, so we have to fix this manually :(

File: pkgs/development/python-modules/tomlkit/default.nix
Comment (Ma27) [inline pkgs/development/python-modules/tomlkit/default.nix]: same as above, Python 3.4 isn't supported on master anymore.

File: pkgs/development/python-modules/pyrsistent/default.nix
Comment (Ma27) [inline pkgs/development/python-modules/pyrsistent/default.nix]: I may be missing something, but why isn't `pytestrunner` part of `checkInputs`?

**Review by Ma27 — COMMENTED** [review]
Regarding the `enableParallelBuilding` thing: I might've missed something while skimming through the thread, but as it's only used for make parallelism (which happens to be the "default" build approach when using `stdenv.mkDerivation`), why is it used here? IIRC `buildPythonPackage` directly invokes `setup.py`.

Another concern I have is that none of these packages has maintainers listed: we already have numerous breaking python packages on Hydra (due to too strict dependency constraints mostly), so I'd like to see as much of these packages as possible having maintainers that can help us keeping this package set as stable as possible.

File: pkgs/development/python-modules/clikit/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/clikit/default.nix]: My thought was that this is the constraint in the build file and it doesn't hurt to leave it here as documentation/just in case.

File: pkgs/development/python-modules/pyrsistent/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/pyrsistent/default.nix]: I think it's because it's necessary even if you're not running tests, just to run `setup.py`. It's in `setup_requires` [here](https://github.com/tobgu/pyrsistent/blob/b3d0c27554595a1c1edc781adefcf029632bdb45/setup.py#L82) and it fails the build when I move it to `checkInputs` if I also set `doCheck = false;`.

**Comment by jbaum98** [comment]
I'm happy to be the maintainer for these packages if that makes sense, but I'll need to add myself to the maintainers list.

**Comment by worldofpeace** [comment]
> https://github.com/NixOS/nixpkgs/pull/53599#discussion_r245964306
> we don't have Python 3.4 on master anymore, so this shouldn't be needed.

> My thought was that this is the constraint in the build file and it doesn't hurt to leave it here as documentation/just in case.

I agree with @jbaum98 even if it's not needed. I tend to try and make the derivation respect the things specified by upstream as close as possible.

Unless they want crazy nonsense that is, then maybe patch it all away :smile:

**Comment by Ma27** [comment]
> I'm happy to be the maintainer for these packages if that makes sense, but I'll need to add myself to the maintainers list.

That shouldn't be a problem I guess. Just add yourself in the first commit in this branch (this is what most new contributors do).

File: pkgs/development/python-modules/pyrsistent/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/pyrsistent/default.nix]: I'm not sure I understand the issue. `pytestrunner` is necessary to build the derivation regardless of whether or not you're running tests, but not to run the executable, so isn't `nativeBuildInputs` exactly where it should go?

Are you saying we should write a patch that makes `pytestrunner` only necessary when you're running tests so that we can move it to `checkInputs`?

Comment (Ma27) [reply pkgs/development/python-modules/pyrsistent/default.nix]: Why do you need `pytestrunner` to build this package? If it's only listed in `setup.py` to ensure it's installed for the tests I'd propose to create a patch which removes that dependency. Then you can list `pytestrunner` in `checkInputs` (where it belongs IMHO) and nothing breaks if the tests need to be disabled in the future.

Or am I missing something?

File: pkgs/development/python-modules/poetry/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/poetry/default.nix]: Python 3.4 is no longer avaliable in Nixpkgs.
@FRidh I wonder why `isPy34` has not yet been removed.

File: pkgs/development/python-modules/poetry/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/poetry/default.nix]:
```suggestion
  postPatch = ''
    subtituteInPlace pyproject.toml --replace "3.0a3" "3.0.0a3"
    subtituteInPlace setup.py --replace "3.0a3" "3.0.0a3"
  '';
```

File: pkgs/development/python-modules/poetry/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/poetry/default.nix]: Please add a comment with the reason.

File: pkgs/development/python-modules/pylev/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/pylev/default.nix]: Why not `fetchPypi`?

File: pkgs/development/python-modules/poetry/default.nix
Comment (Ma27) [reply pkgs/development/python-modules/poetry/default.nix]: this has been discussed [here](https://github.com/NixOS/nixpkgs/pull/53599#discussion_r245964306).

@worldofpeace and @jbaum98 preferred to keep it.

**Comment by jbaum98** [comment]
The pypi tarball for this and others don't include tests. Some of them also don't include a `setup.py` and so need to be built with poetry, so once we bootstrap poetry we actually will be making most of these packages build from GitHub so they can run with tests.

File: pkgs/development/python-modules/pylev/default.nix
Comment (jbaum98) [reply pkgs/development/python-modules/pylev/default.nix]: The pypi tarball for this and others don't include tests. Some of them also don't include a `setup.py` and so need to be built with poetry, so once we bootstrap poetry we actually will be making most of these packages build from GitHub so they can run with tests.

File: pkgs/development/python-modules/pyrsistent/default.nix
Comment (worldofpeace) [reply pkgs/development/python-modules/pyrsistent/default.nix]: > Or am I missing something?

No @Ma27 I think you have it correct. I'll make this change.

pytestrunner even talks about achieving a [conditional requirement](https://github.com/pytest-dev/pytest-runner#conditional-requirement).

File: pkgs/development/python-modules/pylev/default.nix
Comment (worldofpeace) [reply pkgs/development/python-modules/pylev/default.nix]: I think I forgot to instruct that for `pylev` and `pastel` there should be a comment like `No tests in PyPi Tarball`.

File: pkgs/development/python-modules/pyrsistent/default.nix
Comment (worldofpeace) [reply pkgs/development/python-modules/pyrsistent/default.nix]: https://github.com/NixOS/nixpkgs/pull/53599/commits/b8a6d1be56686ffb0a860b5184fd96b025150c05

File: pkgs/development/python-modules/pyrsistent/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/pyrsistent/default.nix]: Did you submit that upstream?

**Review by dotlambda — APPROVED** [review]
The last few commits should be squashed into the appropriate ones. Otherwise, this is fine.

File: pkgs/development/python-modules/pyrsistent/default.nix
Comment (worldofpeace) [reply pkgs/development/python-modules/pyrsistent/default.nix]: Wasn't sure, `pytestrunner` instructs people to do this https://github.com/pytest-dev/pytest-runner#usage

**Review by dotlambda — APPROVED** [review]

**Review by Ma27 — APPROVED** [review]

**Comment by worldofpeace** [comment]
@jbaum98 Thanks for your contribution :fireworks:

File: pkgs/development/python-modules/pyrsistent/default.nix
Comment (CMCDragonkai) [reply pkgs/development/python-modules/pyrsistent/default.nix]: I realised now, that I do that in my Python packages, doing conditional requirement for pytestrunner in the `setup.py`. Although I wonder why is pytestrunner needed in `setup_requires` anyway? Why not just `test_requires`. Maybe some technical limitation in pytestrunner/setup.py.

File: pkgs/development/python-modules/poetry/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/poetry/default.nix]: We usually can't have a Python package depend on a version that's not the default because this can lead to conflicts. Was there a reason we allowed this back then?
And isn't poetry only ever used as an application, i.e. never imported? In that case, we should move it out of `pythonPackages`.

Comment (worldofpeace) [reply pkgs/development/python-modules/poetry/default.nix]: > We usually can't have a Python package depend on a version that's not the default because this can lead to conflicts. Was there a reason we allowed this back then?
And isn't poetry only ever used as an application, i.e. never imported?

Yeah it is weird that I forgot about that, especially since we're always mentioning this :smile: 
I can move this out of `python-modules`.

Comment (dotlambda) [reply pkgs/development/python-modules/poetry/default.nix]: If you're it's never imported, feel free to go ahead. Thank you in advance!

**Comment by adisbladis** [comment]
Fyi I've repackaged `poetry` with `poetry2nix` in https://github.com/adisbladis/poetry2nix/tree/master/pkgs/poetry (1.0.0b5).

This fixes a bunch of issues that this package currently has:
- It doesn't propagate `PYTHONPATH` 
This ensures that things building with poetry as a PEP518(`pyproject.toml`) `build-system` will not get a mix of dependencies from your `pyproject.toml` and nixpkgs `poetry`.
It does this by using a similar "vendoring" approach as the official poetry installer does.
- `build-system` is a first class supported thing
Poetry is injected into the build if it's defined as a `build-system`
- It does not require overriding any modules from `python-packages.nix`
A unique coherent package set is created for every leaf package, thus avoiding conflicts between packages.
- It's dynamic and builds from `pyproject.toml`/`poetry.lock`.
This makes packaging things much easier, a typical Poetry project packaged looks something like:
``` nix
poetry2nix.mkPoetryPackage {
  inherit python;
  pyproject = ./pyproject.toml;
  poetryLock = ./poetry.lock;
  src = fetchFromGitHub {
    owner = "sdispater";
    repo = "poetry";
    rev = "1.0.0b5";
    sha256 = "1l7vd233p9g8sss5cxl6rm6qyf55ahbb9sw7zm3sd5xchcsiyday";
  };
}
```

I intend to get this into nixpkgs as soon as Poetry releases a 1.0.

**Pull request merged**
